File: launcher/compress.py
"""On-the-fly gzip compression for text-based responses.

Compression is applied only when the client sends `Accept-Encoding: gzip`,
the body is compressible, and it is large enough to be worth the CPU.
Static files get an in-memory gzip cache keyed by (mtime, size) so repeated
requests don't re-compress; dynamic responses (JSON, index) are compressed
per request and never cached.
"""

import gzip
import threading

_COMPRESSIBLE = frozenset({
    "text/html",
    "text/css",
    "text/javascript",
    "application/javascript",
    "application/json",
})
_MIN_SIZE = 1024
_LEVEL = 6

_cache = {}
_cache_lock = threading.Lock()


def wants_gzip(accept_encoding):
    return "gzip" in (accept_encoding or "").lower()


def should_compress(content_type, size):
    base = (content_type or "").split(";")[0].strip().lower()
    return size >= _MIN_SIZE and base in _COMPRESSIBLE


def gzip_bytes(body):
    return gzip.compress(body, _LEVEL)


def gzip_static(path, body):
    """gzip `body` with a per-file cache (re-compresses when the file changes)."""
    try:
        st = path.stat()
        stamp = (st.st_mtime_ns, st.st_size)
    except OSError:
        return gzip_bytes(body)
    key = str(path)
    with _cache_lock:
        cached = _cache.get(key)
        if cached and cached[0] == stamp:
            return cached[1]
    data = gzip.compress(body, _LEVEL)
    with _cache_lock:
        _cache[key] = (stamp, data)
    return data

File: internal/compress/compress.go
// Package compress provides on-the-fly gzip for text responses, mirroring
// the Python implementation: only compress when the client accepts gzip,
// the content type is compressible and the body is big enough to be worth
// the CPU. Static files get an in-memory gzip cache keyed by (mtime, size)
// so repeated requests don't re-compress; embedded files use a cache keyed
// by content hash since they never change.
package compress

import (
	"bytes"
	"compress/gzip"
	"hash/fnv"
	"net/textproto"
	"strings"
	"sync"
)

var compressible = map[string]bool{
	"text/html":             true,
	"text/css":              true,
	"text/javascript":       true,
	"application/javascript": true,
	"application/json":      true,
}

const minSize = 1024
const level = 6

// WantsGzip reports whether the Accept-Encoding header includes gzip.
func WantsGzip(acceptEncoding string) bool {
	return strings.Contains(strings.ToLower(acceptEncoding), "gzip")
}

// ShouldCompress reports whether a body of the given type and size is
// worth compressing.
func ShouldCompress(contentType string, size int) bool {
	base := strings.TrimSpace(strings.ToLower(strings.Split(contentType, ";")[0]))
	return size >= minSize && compressible[base]
}

// GzipBytes compresses body at the standard level.
func GzipBytes(body []byte) []byte {
	var buf bytes.Buffer
	w, _ := gzip.NewWriterLevel(&buf, level)
	w.Write(body)
	w.Close()
	return buf.Bytes()
}

type cacheEntry struct {
	key  uint64
	data []byte
}

var (
	cache     = map[string]cacheEntry{}
	cacheLock sync.Mutex
)

// GzipCached compresses body once per key (e.g. "static/js/app.js"),
// re-compressing only when the fingerprint changes. For embedded assets
// the fingerprint is a content hash, so the cache stays valid forever.
func GzipCached(key string, body []byte) []byte {
	h := fnv.New64a()
	h.Write(body)
	stamp := h.Sum64()
	cacheLock.Lock()
	cached, ok := cache[key]
	cacheLock.Unlock()
	if ok && cached.key == stamp {
		return cached.data
	}
	data := GzipBytes(body)
	cacheLock.Lock()
	cache[key] = cacheEntry{key: stamp, data: data}
	cacheLock.Unlock()
	return data
}

// NormalizeType trims parameters ("text/html; charset=utf-8" ->
// "text/html") in the same casing style the Python code compared.
func NormalizeType(contentType string) string {
	return textproto.TrimString(strings.ToLower(strings.Split(contentType, ";")[0]))
}
